bundle/src/main/java/com/adobe/acs/commons/users/impl/EnsureAce.java

```diff
@@ -0,0 +1,281 @@
+     * @param serviceUser
+     *            the Service User
+     * @throws RepositoryException
+     */
+    public void removeAces(ResourceResolver resourceResolver, Authorizable authorizable,
+            AbstractAuthorizable serviceUser) throws RepositoryException {
+        final Session session = resourceResolver.adaptTo(Session.class);
+
+        final JackrabbitAccessControlManager accessControlManager =
+                (JackrabbitAccessControlManager) session.getAccessControlManager();
+        final List<JackrabbitAccessControlList> acls =
+                findAcls(resourceResolver, serviceUser.getPrincipalName(), accessControlManager);
+
+        for (final JackrabbitAccessControlList acl : acls) {
+            final JackrabbitAccessControlEntry[] aces = (JackrabbitAccessControlEntry[]) acl.getAccessControlEntries();
+
+            // Check all the existing ACEs in the ACL
+            for (JackrabbitAccessControlEntry ace : aces) {
+                if (StringUtils.equals(serviceUser.getPrincipalName(), ace.getPrincipal().getName())) {
+                    if (authorizable != null && StringUtils.startsWith(acl.getPath(), authorizable.getPath())) {
+                        // Skip! Don't ensureRemoval ACE's from the system user itself!
+                    } else {
+                        acl.removeAccessControlEntry(ace);
+                    }
+                }
+            }
+
+            accessControlManager.setPolicy(acl.getPath(), acl);
+            log.debug("Removed ACE from ACL at [ {} ] for [ {} ]", acl.getPath(), serviceUser.getPrincipalName());
+        }
+    }
+
+    /**
+     * Locates by query all the ACLs that the principal participates in.
+     *
+     * @param resourceResolver
+     *            the resource resolver to perform the user management
+     * @param principalName
+     *            the principal name
+     * @param accessControlManager
+     *            Jackrabbit access control manager
+     * @return a list of ACLs that principal participates in.
+     * @throws RepositoryException
+     */
+    private List<JackrabbitAccessControlList> findAcls(ResourceResolver resourceResolver, String principalName,
+            JackrabbitAccessControlManager accessControlManager) throws RepositoryException {
+        final Set<String> paths = new HashSet<String>();
+        final List<JackrabbitAccessControlList> acls = new ArrayList<JackrabbitAccessControlList>();
+
+        final Map<String, String> params = new HashMap<String, String>();
+
+        params.put("type", AccessControlConstants.NT_REP_ACE);
+        params.put("property", AccessControlConstants.REP_PRINCIPAL_NAME);
+        params.put("property.value", principalName);
+        params.put("p.limit", "-1");
+
+        final Query query =
+                queryBuilder.createQuery(PredicateGroup.create(params), resourceResolver.adaptTo(Session.class));
+        final Iterator<Resource> resources = query.getResult().getResources();
```

Comment:
QueryBuilder has a RR resource leak - it'd better to write this as:

```
query.getResult().getPaths().steam().forEach(path -> { Resource resource = resourceResolver.getResource(path); ... }
```

Reply:
I don't believe (I could be wrong) I changed that code, just copied it to a new class to make it easier to re-use. either way...

I don't see a `getPaths()` method on the result object, so I'm using `query.getResult().getHits().forEach(hit->{})` hope that works.

Would be curious to know more about the Resource Leak in QB, Is that documented as a known issue anywhere?  As far I can tell the code doesn't "look" like it should cause a leak, unless I'm missing something.

Reply:
yup - was going off memory on this one; youre right - there is not `getPaths()` ... what you suggested (`query.getResult().getHits().forEach(hit->{rr.getResource(hit.getPath()})`) is what I meant! 

The QB issue is logged as an issue via Adobe Support; The issue is a JCR Session object is passed into QB to provided security context, and that JCR Session is then used to open a new RR, and QB never closes this RR ... in large part because its very hard to tell when its safe to close it .. since when its closed since objects of the ilk `query.getResult.getResources()` will become invalid (since their backing RR has been closed).

Not engaging any of the "Sling" methods on QB's results happily avoids the opening on the RR, so its safe (even if its only "safe" by way of a QB implementation detail).